### PR TITLE
Change EmbedLayerNormalization mask index output to optional

### PR DESCRIPTION
**Pull request by kunal-vaishnavi**
### Description
This PR changes an EmbedLayerNormalization node's mask index output to be an optional output if a mask input is not provided.

### Motivation and Context
The documentation for EmbedLayerNormalization states 
```
The last input mask is optional. If mask is provided, mask index (that is position of first 0 in mask, or number of words) will be calculated.
```
However, if the mask input is not provided, the mask index output is still calculated and required.

### CODE REVIEW

File: onnxruntime/core/graph/contrib_ops/shape_inference_functions.cc [hunk 1]
```diff
@@ -8,7 +8,9 @@ namespace onnxruntime {
 namespace contrib {
 void EmbedLayerNormalizationShapeInference(::ONNX_NAMESPACE::InferenceContext& ctx) {
   propagateElemTypeFromInputToOutput(ctx, 2, 0);
-  propagateElemTypeFromInputToOutput(ctx, 0, 1);
+  if (ctx.getNumOutputs() > 1) {
```
Comment (tianleiwu) [inline onnxruntime/core/graph/contrib_ops/shape_inference_functions.cc:11]: This is not correct. There are two optional outputs, so two outputs could be output + mask_index or output + embedding_sum.

File: onnxruntime/core/graph/contrib_ops/shape_inference_functions.cc [hunk 2]
```diff
@@ -97,9 +99,11 @@ void EmbedLayerNormalizationShapeInference(::ONNX_NAMESPACE::InferenceContext& c
...
-  ONNX_NAMESPACE::TensorShapeProto mask_index_shape;
-  *mask_index_shape.add_dim() = input_ids_dims[0];
-  updateOutputShape(ctx, 1, mask_index_shape);
+  if (ctx.getNumOutputs() > 1) {
+    ONNX_NAMESPACE::TensorShapeProto mask_index_shape;
+    *mask_index_shape.add_dim() = input_ids_dims[0];
+    updateOutputShape(ctx, 1, mask_index_shape);
+  }
 
   if (ctx.getNumOutputs() > 2) {
```
Comment (tianleiwu) [inline onnxruntime/core/graph/contrib_ops/shape_inference_functions.cc:108]: May need update logic here if we want to support outputs output + embedding_sum (mask_index is optional)

File: onnxruntime/contrib_ops/cuda/bert/embed_layer_norm.cc [hunk 3]
```diff
@@ -55,18 +56,25 @@ Status EmbedLayerNorm<T>::ComputeInternal(OpKernelContext* context) const {
...
   int sequence_length = static_cast<int>(input_dims[1]);
-  size_t element_size = sizeof(T);
 
   const bool broadcast_position_ids = (nullptr != position_ids && position_ids->Shape()[0] == 1);
 
   return LaunchEmbedLayerNormKernel(
       Stream(context),
           output->MutableData<T>(),
-          mask_index->MutableData<int32_t>(),
+          (mask_index != nullptr) ? mask_index->MutableData<int32_t>() : dummy_mask_index,
```
[the remark this answers is not in the document]
Comment (kunal-vaishnavi) [reply onnxruntime/contrib_ops/cuda/bert/embed_layer_norm.cc]: Fixed

File: onnxruntime/core/graph/contrib_ops/shape_inference_functions.cc (hunk 1 above)
Comment (kunal-vaishnavi) [reply onnxruntime/core/graph/contrib_ops/shape_inference_functions.cc]: Updated this check to see if the mask input shape exists

File: onnxruntime/core/graph/contrib_ops/shape_inference_functions.cc (hunk 2 above)
Comment (kunal-vaishnavi) [reply onnxruntime/core/graph/contrib_ops/shape_inference_functions.cc:108]: Updated this check to see if the mask input shape exists

Comment (tianleiwu) [reply onnxruntime/core/graph/contrib_ops/shape_inference_functions.cc:108]: Could we use ctx.getOutputType(1) == nullptr to check whether the output exists?

**Comment by tianleiwu** [comment]
This is not backward compatible (For example, previously mask_index has shape even when there is no mask input). Maybe add an attribute mask_index_type, and add a unit test.

File: onnxruntime/core/graph/contrib_ops/shape_inference_functions.cc [hunk 4]
```diff
@@ -3,12 +3,17 @@
...
 #include <onnx/defs/shape_inference.h>
+#include <iostream>
 
 namespace onnxruntime {
 namespace contrib {
 void EmbedLayerNormalizationShapeInference(::ONNX_NAMESPACE::InferenceContext& ctx) {
   propagateElemTypeFromInputToOutput(ctx, 2, 0);
-  if (!ctx.getOutputType(1)) {
+  auto mask_index_type = getAttribute(ctx, "mask_index_type", 1);
+  // std::cout << "Num outputs: " << (ctx.getNumOutputs()) << std::endl;
```
Comment (tianleiwu) [inline onnxruntime/core/graph/contrib_ops/shape_inference_functions.cc:13]: Remove commented code.

**Comment by tianleiwu** [comment]
Please look at test error in Windows GPU CI pipeline:
`
Exception during initialization: D:\a\_work\1\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\DmlGraphFusionHelper.cpp(396)\onnxruntime_test_all.exe!00007FF69EEC4AC7: (caller: 00007FF69EEC5352) Exception(1) tid(a20) 80070057 The parameter is incorrect.
`

File: onnxruntime/core/graph/contrib_ops/shape_inference_functions.cc (hunk 4 above)
Comment (kunal-vaishnavi) [reply onnxruntime/core/graph/contrib_ops/shape_inference_functions.cc:13]: I'll remove it once I finish resolving the Windows GPU CI Pipeline error

**Comment by kunal-vaishnavi** [comment]
> Please look at test error in Windows GPU CI pipeline: `Exception during initialization: D:\a\_work\1\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\DmlGraphFusionHelper.cpp(396)\onnxruntime_test_all.exe!00007FF69EEC4AC7: (caller: 00007FF69EEC5352) Exception(1) tid(a20) 80070057 The parameter is incorrect.`

I'm currently investigating this error

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**